File: app/src/main/kotlin/com/foodshare/features/challenges/domain/repository/ChallengeRepository.kt
package com.foodshare.features.challenges.domain.repository

import com.foodshare.features.challenges.domain.model.*

/**
 * Repository interface for challenge operations.
 *
 * SYNC: This mirrors Swift ChallengeRepository protocol
 */
interface ChallengeRepository {
    /**
     * Fetch all published challenges.
     */
    suspend fun getPublishedChallenges(): Result<List<Challenge>>

    /**
     * Fetch challenges for a specific user with their status.
     */
    suspend fun getUserChallenges(userId: String): Result<List<ChallengeWithStatus>>

    /**
     * Fetch user challenges with joined/completed counts.
     */
    suspend fun getUserChallengesWithCounts(userId: String): Result<UserChallengesWithCountsResult>

    /**
     * Fetch a specific challenge with user's status.
     */
    suspend fun getChallenge(id: Int, userId: String): Result<ChallengeWithStatus>

    /**
     * Accept a challenge (mark as joined).
     */
    suspend fun acceptChallenge(challengeId: Int, userId: String): Result<ChallengeActivity>

    /**
     * Mark a challenge as completed.
     */
    suspend fun completeChallenge(challengeId: Int, userId: String): Result<ChallengeActivity>

    /**
     * Reject/skip a challenge.
     */
    suspend fun rejectChallenge(challengeId: Int, userId: String): Result<ChallengeActivity>

    /**
     * Fetch leaderboard for a challenge.
     */
    suspend fun getLeaderboard(challengeId: Int, limit: Int = 10): Result<List<ChallengeLeaderboardEntry>>

    /**
     * Toggle like on a challenge.
     * @return Pair of (isLiked, newLikeCount)
     */
    suspend fun toggleLike(challengeId: Int, userId: String): Result<Pair<Boolean, Int>>

    /**
     * Check if user has liked a challenge.
     */
    suspend fun hasLiked(challengeId: Int, userId: String): Result<Boolean>
}

File: Sources/FoodShare/Features/Challenges/Domain/Repositories/ChallengeRepository.swift
//
//  ChallengeRepository.swift
//  Foodshare
//
//  Repository protocol for challenge operations
//

import Foundation

// MARK: - User Challenges with Counts Result

struct UserChallengesWithCountsResult: Sendable {
    let challenges: [ChallengeWithStatus]
    let joinedCount: Int
    let completedCount: Int

    static let empty = UserChallengesWithCountsResult(
        challenges: [],
        joinedCount: 0,
        completedCount: 0,
    )
}

protocol ChallengeRepository: Sendable {
    // MARK: - Fetch Challenges

    /// Fetch all published challenges
    func fetchPublishedChallenges() async throws -> [Challenge]

    /// Fetch challenges user has accepted
    func fetchUserChallenges(userId: UUID) async throws -> [ChallengeWithStatus]

    /// Fetch challenges user has accepted with pre-computed counts
    func fetchUserChallengesWithCounts(userId: UUID) async throws -> UserChallengesWithCountsResult

    /// Fetch a single challenge with user status
    func fetchChallenge(id: Int, userId: UUID) async throws -> ChallengeWithStatus

    // MARK: - Participation

    /// Accept a challenge
    func acceptChallenge(challengeId: Int, userId: UUID) async throws -> ChallengeActivity

    /// Complete a challenge
    func completeChallenge(challengeId: Int, userId: UUID) async throws -> ChallengeActivity

    /// Reject a challenge
    func rejectChallenge(challengeId: Int, userId: UUID) async throws -> ChallengeActivity

    // MARK: - Leaderboard

    /// Fetch challenge leaderboard (users who completed)
    func fetchLeaderboard(challengeId: Int, limit: Int) async throws -> [ChallengeLeaderboardEntry]

    // MARK: - Likes

    /// Toggle like on a challenge, returns new like state and count
    func toggleChallengeLike(challengeId: Int, profileId: UUID) async throws -> (isLiked: Bool, likeCount: Int)

    /// Check if user has liked a challenge
    func hasLikedChallenge(challengeId: Int, profileId: UUID) async throws -> Bool
}

// MARK: - Leaderboard Entry

struct ChallengeLeaderboardEntry: Codable, Identifiable, Sendable {
    let id: UUID
    let nickname: String
    let avatarUrl: String?
    let rank: Int
    let isCompleted: Bool

    enum CodingKeys: String, CodingKey {
        case id
        case nickname
        case avatarUrl = "avatar_url"
        case rank
        case isCompleted = "is_completed"
    }
}

#if DEBUG

    extension ChallengeLeaderboardEntry {
        static func fixture(
            id: UUID = UUID(),
            nickname: String = "FoodHero",
            avatarUrl: String? = nil,
            rank: Int = 1,
            isCompleted: Bool = true,
        ) -> ChallengeLeaderboardEntry {
            ChallengeLeaderboardEntry(
                id: id,
                nickname: nickname,
                avatarUrl: avatarUrl,
                rank: rank,
                isCompleted: isCompleted,
            )
        }
    }

#endif
